examples/triangle_intersection/triangle_intersection.cpp

```diff
@@ -0,0 +1,387 @@
+  {
+    return _mappings(i);
+  }
+
+private:
+  Kokkos::View<ArborX::ExperimentalHyperGeometry::Triangle<2> *, MemorySpace>
+      _triangles;
+  Kokkos::View<Mapping *, MemorySpace> _mappings;
+};
+
+// For creating the bounding volume hierarchy given a Triangles object, we
+// need to define the memory space, how to get the total number of objects,
+// and how to access a specific box. Since there are corresponding functions in
+// the Triangles class, we just resort to them.
+template <typename MemorySpace>
+struct ArborX::AccessTraits<Triangles<MemorySpace>, ArborX::PrimitivesTag>
+{
+  using memory_space = MemorySpace;
+
+  static KOKKOS_FUNCTION int size(Triangles<MemorySpace> const &triangles)
+  {
+    return triangles.size();
+  }
+
+  static KOKKOS_FUNCTION auto get(Triangles<MemorySpace> const &triangles,
+                                  int i)
+  {
+    auto const &triangle = triangles(i);
+    ArborX::ExperimentalHyperGeometry::Box<2> box{};
+    box += triangle.a;
+    box += triangle.b;
+    box += triangle.c;
+    return box;
+  }
+};
+
+template <typename MemorySpace>
+struct ArborX::AccessTraits<Points<MemorySpace>, ArborX::PredicatesTag>
+{
+  using memory_space = MemorySpace;
+  static KOKKOS_FUNCTION int size(Points<MemorySpace> const &points)
+  {
+    return points.size();
+  }
+  static KOKKOS_FUNCTION auto get(Points<MemorySpace> const &points, int i)
+  {
+    return ArborX::attach(ArborX::intersects(points(i)), i);
+  }
+};
+
+template <typename MemorySpace>
+class TriangleIntersectionCallback
+{
+public:
+  TriangleIntersectionCallback(
+      Triangles<MemorySpace> triangles,
+      Kokkos::View<int *, MemorySpace> offsets,
+      Kokkos::View<ArborX::Point *, MemorySpace> coefficients)
+      : _triangles(triangles)
+      , _offsets(offsets)
+      , _coefficients(coefficients)
+  {}
+
+  // The search tree consists entirely of boxes, although the primitives are
+  // triangles. Thus, a detected collision doesn't mean that the point is
+  // actuallt inside the wrapped triangle and we have to check that here. This
+  // also gives us the opportunity to store the barycentric coordinates in case
+  // there is an intersection. Since the triangles don't overlap in this
+  // example, there is at most one triangle that contains a given point and we
+  // can abort the search early when we found a match.
+  template <typename Query, typename Primitive>
+  KOKKOS_FUNCTION auto operator()(Query const &query,
+                                  Primitive const &primitive) const
+  {
+    ArborX::ExperimentalHyperGeometry::Point<2> const &point =
+        getGeometry(getPredicate(query));
+    auto query_index = ArborX::getData(query);
+
+    auto const coeffs =
+        _triangles.get_mapping(primitive.index).get_coeff(point);
+    bool intersects = coeffs[0] >= 0 && coeffs[1] >= 0 && coeffs[2] >= 0;
+
+    if (intersects)
+    {
+      _offsets(query_index) = primitive.index;
+      _coefficients(query_index) = coeffs;
```

Comment:
Why do we store coefficients? Is that because the problem we consider is interpolation? I don't remember what XGC or the like want.

Reply:
Yes. In my opinion, this is one of the points of this example.